Ddci phase1

jasmin-j opened the pull request:
I have implemented the support for the Digital Devices CI interfaces (see also issue #4477). Currently only one service can use the dedicated CI at a time. There is no MCD (Multi Channel Decoding) or MTD (Multi Tuner Decoding) implemented.

It is 95% working, but sometimes the CAM doesn't decrypt. This is no problem of the new code, but of the PID selection for CAMs in TVH general, I guess. 

The DD CI code is 100% finished and works, if the PIDs are set properly by TVH. This has been already tested by me and another TVH user.

perexg commented:
There are some things which should be changed, but I'll do it.

- dynamic PID subscription (subscribe only ECM PIDs for active CAID), tjhe current code subscribes all ECM PIDs which is not optimal
- pass always ECM PIDs to descramble_descramble() - to avoid the special conditions for those 'external' decoders in future, the filter in ts_recv_packet0() is enough
- with the above change, we may move ECM handling from tables to descramble_descramble() too (softcam)
- use 'dr->dr_descramble' callback for standard dvbcam (remove dr->dr_external)

and probably others..

perexg commented:
Merged with a ton of my changes (previous comment). Please, test.

herrnst commented:
@perexg @jasmin-j 
Even with the additional changes, the assignment issue remains. The neccessary (all CA) PIDs are not reaching the CAM, I guess caused due to the CAM not being assigned correctly - in Jasmin's original code I was able to get things working by modifying https://github.com/tvheadend/tvheadend/blob/9d1f7dbf1ddb1a9c77b69fa8928f25ad83ca16bd/src/input/mpegts/mpegts_input.c#L730 to "is_ddci = 1", that ifdef has gone now, though.

debug/trace=en50221,ddci,caclient,descrambler,dvbcam (still) results in this log output when switching to an encrypted channel: https://pastebin.com/y6Y9ux7F

The CAM (AlphaCrypt running One4All) though is (obviously) supposed to decrypt the stream, or at least tell me (after querying the inserted card) via MMI that one doesn't have permission to decrypt the service.

If you need more debug, someone to test patches or whatever, just let me know.

jasmin-j commented:
> Even with the additional changes, the assignment issue remains. The neccessary (all CA) PIDs are not
> reaching the CAM

This is true. The reason is, that the nobody is executing "descrambler_open_emm" and also the CAT PIDs are not subscribed correctly. I started to implement this with a new "dvbcam_caid_update" function, but although the PIDs are opened by mpegts and linuxdvb, they are not sent to the CAM. The even doesn't come out of the dvrX device.
The old code used "mpegts_input_cat_pass_callback" to decode the CAT and subscribe the PIDs. This function is still called, but the CAT PIDs are not subscribed.

Also the CAT is not sent to the CAM, but the PAT. The CAM requires the CAT, but not the PAT.
I changed that in "dvbcam_service_start" by calling "mpegts_input_open_service_pid" with "DVB_CAT_PID", but I still doesn't get a picture. And this is, because the CAT PIDs are not subscribed, I guess.
And the CAM needs also the EMM PIDs. I added some code to "dvbcam_caid_update", but this seems to be the wrong function. I have to debug this or maybe @perexg can point me in the right direction.

> I guess caused due to the CAM not being assigned correctly

No, the assignment of the DD CI to the service works. See above for the root cause.
Do to the recent changes from @perexg the whole PID mapping is done a bit different (better) and the PIDs are attached by "dvbcam_caid_update".

BR,
Jasmin

Trujulu commented:
Decription is failing on oscam (dvbapi) for me for some channels now... There is something bad with new change...

herrnst commented:
FWIW: OSCam via newcamd (cba to change everything to networked dvbapi) still works/decrypts fine with current HEAD.

perexg commented:
@jasmin-j  , @herrnst : Could you test latest? The CAT issue should be fixed with 8a61d01d15c1a4fa1232c3b79bd01aa5ef1ad3ec .

Trujulu commented:
@hernnst I will retest monday morning. Maybe oscam fault.

perexg commented:
@Trujulu : Create a new issue with --trace descrambler,capmt , but the changes should not break anything.

perexg commented:
@jasmin-j, @herrnst : I forgot to add CAT PID to the streaming data : https://github.com/tvheadend/tvheadend/commit/a8fcc4cb1d803fccfa62479b04ac55b250a1dd1f

herrnst commented:
@perexg Current HEAD (23741fb) causes a deadlock when tuning to an encrypted service with the CICAM enabled, last things logged (components as mentioned above):

    Nov 17 19:06:52 dvbdev tvheadend[5381]: mpegts: 354MHz in Unitymedia NRW - tuning on ST STV0367 DDB DVB-C/T : DVB-C #0
    Nov 17 19:06:52 dvbdev tvheadend[5381]: dvbcam: start service 0x55f1b694e9a0
    Nov 17 19:06:52 dvbdev tvheadend[5381]: ddci: CAM ci0 assigned to 0x55f1b694e9a0
    Nov 17 19:06:52 dvbdev tvheadend[5381]: descrambler: dvbcam-0-0-1838: key state changed from INIT to READY for "Unitymedia NRW/354MHz/Fox HD"
    Nov 17 19:06:52 dvbdev tvheadend[5381]: descrambler: service "Unitymedia NRW/354MHz/Fox HD": 0 descramblers (0 ok 0 failed)`

With the background/idle/initial EPG scanning active (disabled all of them inbetween to get a clean log, thus timestamp is "in the past"), the log gets then filled with

    Nov 17 19:02:42 dvbdev tvheadend[4993]: mpegts: too much queued table input data (over 2MB), discarding new

herrnst commented:
@perexg a8fcc4c still deadlocks, last things getting logged are the same as posted six minutes earlier.

herrnst commented:
"gcc -Og -g" build, gdb bt full and thread apply all bt: https://pastebin.com/yWLJ8Cz4

perexg commented:
@herrnst : https://github.com/tvheadend/tvheadend/commit/76b77e0a82a086b13dce3e124df89b9ffc965ab4

herrnst commented:
@perexg Thanks, still no luck though. The deadlock is gone, but still the CAM just sends back the encrypted packets, ie:

    Nov 18 10:13:15 dvbdev tvheadend[4542]: ddci: CAM ci0 recv rd(CAM ->):118272, wr:118272
    Nov 18 10:13:15 dvbdev tvheadend[4542]: ddci: CAM ci0 got 114309 scrambled packets from CAM

so it looks like it still doesn't get the CA PIDs (not sure how to verify this ATM but the effect remains the same).

perexg commented:
"--trace mpegts" - https://tvheadend.org/projects/tvheadend/wiki/Traces

herrnst commented:
@perexg https://pastebin.com/6eQEzRbF

PID list on that service (paste from WebUI / service info):

    Index	PID	Type	Language	Details
     	0x01ff /   511	PCR	 	 
     	0x0061 /    97	PMT	 	 
    1	0x01ff /   511	H264	 	 
    2	0x0203 /   515	AC3	ger	 
    3	0x0204 /   516	AC3	eng	 
    4	0x195e /  6494	CA	 	CAIDS: 1850:000000
    5	0x1b5e /  7006	CA	 	CAIDS: 1835:000000
    6	0x1c5e /  7262	CA	 	CAIDS: 1838:000000
    7	0x1d5e /  7518	CA	 	CAIDS: 1831:000000
    8	0x1e5e /  7774	CA	 	CAIDS: 098e:000000
    9	0x1500 /  5376	CA	 	CAIDS:
    10	0x1502 /  5378	CA	 	CAIDS:
    11	0x1503 /  5379	CA	 	CAIDS:
    12	0x1504 /  5380	CA	 	CAIDS:
    13	0x1505 /  5381	CA	 	CAIDS:
    14	0x1506 /  5382	CA	 	CAIDS:
    15	0x1507 /  5383	CA	 	CAIDS:

The UM02 card in the AC is Nagra/1838.

jasmin-j commented:
I added some code to subscribe the PIDs (and debugging prints).
Look [here](https://github.com/jasmin-j/tvheadend/commit/ce3ce8f0402e58cee1b0d1e0717bdcdbe58c270d)  for the current state.
All the PIDs are now sent to the CAM, but still no picture :(
I can see EMMs, but no ECMs. The latter are required for descrambling.
I will continue later (have to work today).
@perexg: If you continue, please send me an eMail, so we will not work both that issue.

jasmin-j commented:
Continued by #1034.
